File: 백준/Gold/2493. 탑/탑.py
import sys
N = int(sys.stdin.readline().rstrip())

tops = list(map(int,sys.stdin.readline().split()))

stack = []
for i in range(N):
    while len(stack) != 0 and tops[stack[-1]] < tops[i]:
        stack.pop()

    if len(stack) == 0:
        print(0, end=' ')
    else:
        print(stack[-1] + 1, end=' ')

    stack.append(i)

File: 백준/Gold/2493. 탑/탑.java
import java.io.*;
import java.util.*;

class Main {
   public static void main(String[] args) throws IOException {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

      int N = Integer.parseInt(br.readLine());

      int[] top = new int[N];
      String[] st = br.readLine().split(" ");
      for (int i = 0; i < N; i++) {
        top[i] = Integer.parseInt(st[i]);
      }

      Deque<Integer> stack = new ArrayDeque<>();
      int[] answer = new int[N];

      for (int i = N-1; i >= 0; i--) {
        
        while (!stack.isEmpty() && top[stack.peek()] <= top[i]) {
          int pop = stack.poll();
          answer[pop] = i+1;
        }

        stack.addFirst(i);
      }

      StringBuilder sb = new StringBuilder();
      for (int i = 0; i < N; i++) {
        sb.append(answer[i]).append(" ");
      }

      System.out.println(sb.toString().trim());
   }

}